src/MultiMap.purs

```diff
@@ -0,0 +1,43 @@
+module MultiMap
+  ( MultiMap
+  , empty
+  , singleton
+  , insert
+  , delete
+  , lookup
+  ) where
+
+import Control.Alt (map)
+import Control.Bind ((>>=))
+import Control.Category ((>>>))
+import Data.Array as Array
+import Data.Bounded (class Ord)
+import Data.Function (($))
+import Data.Map as Map
+import Data.Maybe (Maybe)
+import Data.Monoid ((<>))
+
+-- | A specific-use MultiMap where values for the same key are stored in FIFO container
+--   implemented via Array
+newtype MultiMap (k :: Type) (v :: Type) = MultiMap (Map.Map k (Array v))
```

Comment:
`List` could be used instead of array for faster inserts. `insert` should append to the front of the list.

Reply:
I've run some benchmarks it has the same performance in `insert` and some operations will be faster with `Array`